impeller/renderer/backend/vulkan/pipeline_cache_data_vk.cc

```diff
@@ -0,0 +1,116 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "impeller/renderer/backend/vulkan/pipeline_cache_data_vk.h"
+
+#include "flutter/fml/file.h"
+#include "impeller/base/allocation.h"
+#include "impeller/base/validation.h"
+
+namespace impeller {
+
+static constexpr const char* kPipelineCacheFileName =
+    "flutter.impeller.vkcache";
+
+bool PipelineCacheDataPersist(const fml::UniqueFD& cache_directory,
+                              const VkPhysicalDeviceProperties& props,
+                              const vk::UniquePipelineCache& cache) {
+  if (!cache_directory.is_valid()) {
+    return false;
+  }
+  size_t data_size = 0u;
+  if (cache.getOwner().getPipelineCacheData(*cache, &data_size, nullptr) !=
+      vk::Result::eSuccess) {
+    VALIDATION_LOG << "Could not fetch pipeline cache size.";
+    return false;
+  }
+  if (data_size == 0u) {
+    return true;
+  }
+  auto allocation = std::make_shared<Allocation>();
+  if (!allocation->Truncate(Bytes{sizeof(PipelineCacheHeaderVK) + data_size})) {
+    VALIDATION_LOG << "Could not allocate pipeline cache data staging buffer.";
+    return false;
+  }
+  const auto header = PipelineCacheHeaderVK{props, data_size};
+  std::memcpy(allocation->GetBuffer(), &header, sizeof(header));
+  if (cache.getOwner().getPipelineCacheData(
+          *cache, &data_size, allocation->GetBuffer() + sizeof(header)) !=
+      vk::Result::eSuccess) {
+    VALIDATION_LOG << "Could not copy pipeline cache data.";
+    return false;
+  }
+
+  auto allocation_mapping = CreateMappingFromAllocation(allocation);
+  if (!allocation_mapping) {
```

Comment:
It would be nice if we had a unique_ptr variant for when we don't need shared ownership like here. Not blocking.

Reply:
Once we get concepts, I'd like to apply one for any pointer like thing that can be moved. So it would work for both.